src/Features/Hud/VelocityGraph.cpp

```diff
@@ -0,0 +1,140 @@
+#include "VelocityGraph.hpp"
+
+#include "Event.hpp"
+#include "Modules/Client.hpp"
+#include "Modules/Engine.hpp"
+#include "Modules/Scheme.hpp"
+#include "Modules/Surface.hpp"
+#include "Variable.hpp"
+
+Variable sar_hud_velocitygraph("sar_hud_velocitygraph", "0", "Shows velocity graph.\n");
+Variable sar_hud_velocitygraph_font_index("sar_hud_velocitygraph_font_index", "1", 0, "Font index of velocity graph.\n");
+
+VelocityGraph velocityGraph;
+
+VelocityGraph::VelocityGraph()
+	: Hud(HudType_InGame | HudType_Paused | HudType_Menu, true) {
+}
+bool VelocityGraph::ShouldDraw() {
+	return sar_hud_velocitygraph.GetBool() && Hud::ShouldDraw();
+}
+void VelocityGraph::GatherData(int slot) {
+	auto player = client->GetPlayer(slot + 1);
+
+	if (!player) {
+		if (!data.empty())
+			data.clear();
+
+		return;
+	}
+
+	if (data.size() > 500)
+		data.pop_back();
+
+	velocity_data_t current_data;
+
+	auto vel = client->GetLocalVelocity(player);
+	int speed = vel.Length2D();
+
+	unsigned int groundHandle = *(unsigned int *)((uintptr_t)player + Offsets::C_m_hGroundEntity);
+	bool on_ground = groundHandle != 0xFFFFFFFF;
+
+	current_data.speed = speed;
+	current_data.on_ground = on_ground;
+
+	data.insert(data.begin(), current_data);
+}
+
+static int last_vel = 0;
+static int tick_prev = 0;
+
+static int take_off = 0;
+static int take_off_time = 0;
+
+void VelocityGraph::Paint(int slot) {
+	GatherData(slot);
```

Comment:
This gathers data every frame, which means the behaviour of this HUD depends on framerate. You should gather data every tick instead, using a PRE_TICK event handler or similar